pkg/modinstaller/mod_installer.go

```diff
@@ -216,18 +228,169 @@ func (i *ModInstaller) InstallWorkspaceDependencies(ctx context.Context) (err er
+	changes := ChangeSet{}
+	for _, requiredMod := range oldRequire.Mods {
+		modInUpdated := newRequire.GetModDependency(requiredMod.Name)
+		if modInUpdated == nil {
+			continue
+		}
+		if modInUpdated.VersionString != requiredMod.VersionString {
+			changes = append(changes, &Change{
+				Operation:   Replace,
+				OffsetStart: requiredMod.VersionRange.Start.Byte,
+				OffsetEnd:   requiredMod.VersionRange.End.Byte,
+				Content:     []byte(fmt.Sprintf("version = \"%s\"", modInUpdated.VersionString)),
+			})
+		}
+	}
+	return changes
+}
+
+// loads the contents of the mod.sp file and wraps it with a thin wrapper
+// to assist in byte sequence manipulation
+func (i *ModInstaller) loadModFileBytes() (*ByteSequence, error) {
+	modFileBytes, err := os.ReadFile(filepaths.ModFilePath(i.workspaceMod.ModPath))
+	if err != nil {
+		return nil, err
+	}
+	return NewByteSequence(modFileBytes), nil
+}
+
+// creates a new "mod" block which can be written as part of the "require" block in mod.sp
+func (i *ModInstaller) createNewModRequireBlock(modVersion *modconfig.ModVersionConstraint) *hclwrite.Block {
+	modRequireBlock := hclwrite.NewBlock("mod", []string{modVersion.Name})
+	modRequireBlock.Body().SetAttributeValue("version", cty.StringVal(modVersion.VersionString))
+	return modRequireBlock
+}
+
+// updates the 'require' block in 'mod.sp'
+func (i *ModInstaller) updateModFile() error {
+	contents, err := i.loadModFileBytes()
+	if err != nil {
+		return err
+	}
+
+	changes := NewChangeSet()
+
+	oldRequire := i.oldRequire
+	newRequire := i.workspaceMod.Require
+
+	// fill these requires in with empty requires
+	// so that we don't have to do nil checks everywhere
+	// from here on out - if it's empty - it's nil
+
+	if oldRequire == nil {
+		// use an empty require as the old requirements
+		oldRequire = modconfig.NewRequire()
+	}
+	if newRequire == nil {
+		// use a stub require instance
+		newRequire = modconfig.NewRequire()
+	}
+
+	if newRequire.Empty() && !oldRequire.Empty() {
+		// if the new set of requires is empty - just remove the require block
+		changes = append(changes, &Change{
+			Operation:   Delete,
+			OffsetStart: oldRequire.DeclRange.Start.Byte,
+			OffsetEnd:   oldRequire.BodyRange.End.Byte,
+		})
+	} else if oldRequire.Empty() && !newRequire.Empty() {
+		// if the new require is not empty, but the old one is
+		// add a new require block with the new stuff
+		// by generating the HCL string that goes in
+		f := hclwrite.NewEmptyFile()
+		requireBlock := f.Body().AppendNewBlock("require", nil)
+		for _, mvc := range newRequire.Mods {
+			newBlock := i.createNewModRequireBlock(mvc)
+			requireBlock.Body().AppendBlock(newBlock)
+		}
+		changes = append(changes, &Change{
+			Operation:   Insert,
+			OffsetStart: i.workspaceMod.DeclRange.End.Byte - 1,
+			Content:     f.Bytes(),
+		})
+	} else if !i.oldRequire.Empty() && !newRequire.Empty() {
```

Comment:
could this just be 
```
else {
// both empty
```

is this logically correct?

Reply:
ok - in general I feel the logic for this section is a bit confusing/messy - lets discuss